### PR TITLE
gitlab_user: add support for identity provider

### PR DESCRIPTION
**Pull request by LennertMertens**
##### SUMMARY

This change includes the functionality to add and identity provider and the external UID that is required to use an external provider (e.g. Keycloak, google_oauth2, ...).
With this change I also included the option `reset_pasword` because either `password` or `reset_password` is required by the GitLab API. If you use an external identity provider you don't want/need to set a password so that is why the `reset_password` was mandatory for this change.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

- `gitlab_user`

##### ADDITIONAL INFORMATION

This module changes are for people who want to connect their external identity provider with a GitLab user.
You can add users with or without external provider now.

```paste below
TASK [Create GitLab user] ***********************************************************************************************************************************************************************
changed: [localhost] => {"changed": true, "msg": "Successfully created or updated the user testuser123", "user": {"avatar_url": "https://secure.gravatar.com/avatar/59bca95eaaceda1e2d33fa725775d122?s=80&d=identicon", "bio": "", "bio_html": "", "can_create_group": true, "can_create_project": true, "color_scheme_id": 1, "confirmed_at": null, "created_at": "2021-06-02T11:18:49.522+02:00", "current_sign_in_at": null, "email": "lennert@example.be", "external": false, "id": 74, "identities": [{"extern_uid": "f232f...", "provider": "Keycloak"}], "is_admin": false, "job_title": "", "last_activity_on": null, "last_sign_in_at": null, "linkedin": "", "location": null, "name": "Test user", "note": null, "organization": null, "private_profile": false, "projects_limit": 100, "public_email": "", "skype": "", "state": "active", "theme_id": 2, "twitter": "", "two_factor_enabled": false, "username": "testuser123", "web_url": "https://git.example.com/testuser123", "website_url": "", "work_information": null}}
```

### CODE REVIEW

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:60]: This is needed for all new options:
```suggestion
    type: bool
    version_added: 3.2.0
```

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:59]:
```suggestion
    default: false
```
Always use `true`/`false` for YAML booleans.

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:259]: This way, the next diff will be nicer.
```suggestion
                'extern_uid': options['extern_uid'],
            })
```

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:542]: This boolean value is **not** a secret.
```suggestion
        reset_password=dict(type='bool', default=False, no_log=False),
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution. Can you please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)? Also I've added a first few comments below:

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:27]: How is @stgrace related to this PR?

Also please note that adding both names to the author list means that you sign up as maintainers for this module.

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:122]: `provider` and `extern_uid` are also new options, so they need `version_added` as well.

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:530]: Why not:
```suggestion
    for key, value in list(arguments.items()):
```
I think that's easier to read.

File: changelogs/fragments/2691-gitlab_user-support-identity-provider.yml
Comment (felixfontein) [inline changelogs/fragments/2691-gitlab_user-support-identity-provider.yml:3]:
```suggestion
  - "gitlab_user - add functionality for adding an external identity provider to a GitLab user (https://github.com/ansible-collections/community.general/pull/2691)."
```

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (stgrace) [reply plugins/modules/source_control/gitlab/gitlab_user.py:27]: I helped my colleague Lennert with this feature, and I am willing to be a maintainer.
We started off using it in a private repo and to contribute it back to here the git history got altered a little bit

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:433]: Does this make sense? Everything in `arguments` has a `dict` as a value, so this should be a no-op.

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:438]: (`None` values are handled here.)

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:443]: Do I interpret it correctly that either none of these two module options must be provided, or both? If yes, this must be mentioned in the options' documentation, and `AnsibleModule()` needs `required_together=[('provider', 'extern_uid')]`.

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:119]: While looking at the code: why not call this `extern_provider`, so it's more obvious that this and `extern_uid` belong together?

Also from the code, it seems that the API allows a list of these. Would it make sense to allow a list of external logins here as well?

Comment (stgrace) [reply plugins/modules/source_control/gitlab/gitlab_user.py:119]: > While looking at the code: why not call this extern_provider, so it's more obvious that this and extern_uid belong together?

It is named after the parameter used in the API call, which made more sense to us

> Also from the code, it seems that the API allows a list of these. Would it make sense to allow a list of external logins here as well?

When listing/getting users, external identities are shown as a list, but in user creation and modification it is added as in this module (ie with `provider` and `extern_uid`)
https://docs.gitlab.com/ee/api/users.html#user-creation
https://docs.gitlab.com/ee/api/users.html#user-modification

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (stgrace) [reply plugins/modules/source_control/gitlab/gitlab_user.py:443]: Yes that is correct

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [reply plugins/modules/source_control/gitlab/gitlab_user.py:119]: > It is named after the parameter used in the API call, which made more sense to us

Since GitLab API parameter naming is not very consistent, I would prefer better naming in the module which makes it easier for users to see which parameters belong together.

> When listing/getting users, external identities are shown as a list, but in user creation and modification it is added as in this module

The API looks ... pretty basic here :) I'm not sure that it's a good idea to model the module after the API in this case. It looks to me that the API was somewhat designed under the assumption there is only one provider, and that assumption isn't true anymore. Ansible modules usually give users some more abstractions from such an API (or at least, they should IMO).

How about having a list of external identities instead, together with a boolean parameter whether to only add missing ones, or also remove ones that aren't present, and making the module properly manage that list for the user? I think that would be a much better UI for this module.

Comment (felixfontein) [reply plugins/modules/source_control/gitlab/gitlab_user.py:119]: (Not adding the removal functionality and that boolean parameter in this PR would be fine for me, that can be added later if someone wants to have that functionality, but allowing to add multiple ones is something I would really prefer.)

Comment (stgrace) [reply plugins/modules/source_control/gitlab/gitlab_user.py:119]: I think it's fine to add it as a list, only thing I am worried about is when adding/changing multiple identities, you have to make multiple API calls, and when one goes wrong either half of the config is applied or we would have to add a rollback mechanism to revert the changes

Comment (felixfontein) [reply plugins/modules/source_control/gitlab/gitlab_user.py:119]: That's true, but that's a problem with every API that does not support transactions and does not have "magic" calls which happen to do everything you want to do in one step. I think it's acceptable if only parts of the changes are applied (I would do removing of login providers as the last step to avoid the situation that all have been removed and adding a new one fails).

Comment (stgrace) [reply plugins/modules/source_control/gitlab/gitlab_user.py]: Changes have been added, you can now add multiple identities at once and overwrite them with a boolean

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:129]:
```suggestion
    default: false
```

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:461]:
```suggestion
        identities = None
        if 'identities' in arguments:
```

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:582]:
```suggestion
        identities=dict(type='list', elements='dict', options=dict(...)),
```
(Need to fill in the spec for the elements.)

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:130]:
```suggestion
    version_added: 3.3.0
    suboptions:
      ...
```
Add documentation of dict entrie here.

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (stgrace) [reply plugins/modules/source_control/gitlab/gitlab_user.py:470]: identities is an empty list when not specified

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (stgrace) [reply plugins/modules/source_control/gitlab/gitlab_user.py]: should be fixed

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (stgrace) [reply plugins/modules/source_control/gitlab/gitlab_user.py:139]: should be fixed

**Comment by stgrace** [comment]
Everything except this https://github.com/ansible-collections/community.general/pull/2691#discussion_r649895159 should be ok now @felixfontein

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [reply plugins/modules/source_control/gitlab/gitlab_user.py:138]: All YAML booleans will eventually be changed to `true`/`false`, right now we avoid adding new other instances.

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:61]:
```suggestion
    version_added: 3.3.0
```

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:131]:
```suggestion
          - User ID for external identity.
```

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:292]:
```suggestion
                    'identities': {'value': options['identities']},
```

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:301]:
```suggestion
                    'overwrite_identities': {'value': options['overwrite_identities']},
```

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:499]: This must not be called in check mode.

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:511]: Same here.

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:593]:
```suggestion
        overwrite_identities=dict(type='bool', default=False),
```

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:492]: You need to return `True` only when non-checkmode would also do. It's best to simply skip the `user.save()` step when `self._module.check_mode` is truish.

**Comment by stgrace** [comment]
https://github.com/ansible-collections/community.general/pull/2691#discussion_r651659803
This should be fixed now

File: changelogs/fragments/2691-gitlab_user-support-identity-provider.yml
Comment (felixfontein) [inline changelogs/fragments/2691-gitlab_user-support-identity-provider.yml:3]: I think the other two features are missing here:
```suggestion
  - "gitlab_user - specifying a password is no longer necessary (https://github.com/ansible-collections/community.general/pull/2691)."
  - "gitlab_user - allow to reset an existing password with the new ``reset_password`` option (https://github.com/ansible-collections/community.general/pull/2691)."
  - "gitlab_user - add functionality for adding external identity providers to a GitLab user (https://github.com/ansible-collections/community.general/pull/2691)."
```

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:121]:
```suggestion
      - List of identities to be added/updated for this user.
      - To remove all other identities from this user, set I(overwrite_identities=true).
```

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:136]:
```suggestion
      - Overwrite identities with identities added in this module.
      - This means that all identities that the user has and that are not listed in I(identities) are removed from the user.
      - This is only done if a list is provided for I(identities). To remove all identities, provide an empty list.
```

**Review by felixfontein — COMMENTED** [review]
Code looks good to me, I've only added some suggestions to improve the documentation.

**Comment by felixfontein** [comment]
@Lunik @Shaps @dj-wasabi @marwatk @metanovii @scodeman @waheedi @zanssa would be glad if someone else could also take a look at this!

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (russoz) [inline plugins/modules/source_control/gitlab/gitlab_user.py:58]:
```suggestion
      - Whether the user can change its password or not.
```
Typo

File: plugins/modules/source_control/gitlab/gitlab_user.py
Comment (russoz) [inline plugins/modules/source_control/gitlab/gitlab_user.py:129]:
```suggestion
      external_uid:
```
Perhaps clearer as `external` ?

Comment (aminvakil) [reply plugins/modules/source_control/gitlab/gitlab_user.py:130]: There has also been some related discussions about naming here: https://github.com/ansible-collections/community.general/pull/2691#discussion_r645117945

Comment (felixfontein) [reply plugins/modules/source_control/gitlab/gitlab_user.py:130]: I'm slightly in favor of renaming, but not that much that I would request a change. @stgrace what do you think?

Comment (felixfontein) [reply plugins/modules/source_control/gitlab/gitlab_user.py:130]: Let's keep this name for now. If someone wants to change it, please create a follow-up PR (must get merged before 3.3.0 is released, i.e. in ~a week).

**Comment by felixfontein** [comment]
I'm going to merge this now so we can proceed with #2824 (which will have conflicts once this is merged).

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@LennertMertens @stgrace thanks for your work on this!
@aminvakil @russoz thanks for reviewing!

**Pull request merged**
